drivers/platform/aducm3029/pwm.c

```diff
@@ -0,0 +1,300 @@
+#include "error.h"
+#include <drivers/tmr/adi_tmr.h>
+#include <drivers/pwr/adi_pwr.h>
+
+#define NS_PER_SEC 1000000000
+
+/******************************************************************************/
+/************************ Functions Declarations ******************************/
+/******************************************************************************/
+
+static int32_t config_clock_init(ADI_TMR_CONFIG *cfg, uint32_t period,
+				 uint32_t duty, uint32_t *match)
+{
+	float		base_time;
+	float		aux;
+	uint32_t	clk_freq;
+	int32_t		ret;
+	/*
+	 * ADI_TMR_CLOCK_HFOSC (26MHz) &&  ADI_TMR_PRESCALER_1
+	 * -> base_period = 38nS -> max_period = 2,490330 ms
+	 * ADI_TMR_CLOCK_HFOSC (26MHz) &&  ADI_TMR_PRESCALER_16
+	 * -> base_period = 615nS -> max_period = 40,304025 ms
+	 * ADI_TMR_CLOCK_HFOSC (26MHz) &&  ADI_TMR_PRESCALER_64
+	 * -> base_period = 2461nS -> max_period = 161,281635 ms
+	 * ADI_TMR_CLOCK_HFOSC (26MHz) &&  ADI_TMR_PRESCALER_256
+	 * -> base_period = 9846nS -> max_period = 645,257610 ms
+	 *
+	 * ADI_TMR_CLOCK_LFOSC (32KHz) &&  ADI_TMR_PRESCALER_1
+	 * -> base_period = 31,25uS -> max_period = 2,04796875 s
+	 * ADI_TMR_CLOCK_LFOSC (32KHz) &&  ADI_TMR_PRESCALER_16
+	 * -> base_period = 500uS -> max_period = 32,767500 s
+	 * ADI_TMR_CLOCK_LFOSC (32KHz) &&  ADI_TMR_PRESCALER_64
+	 * -> base_period = 2mS -> max_period = 131,070 s
+	 * ADI_TMR_CLOCK_LFOSC (32KHz) &&  ADI_TMR_PRESCALER_256
+	 * -> base_period = 8mS -> max_period = 524,280 s
+	 */
+
+	/* Use HFOSC because in general frequency
+	 * for pwms are lower than 645 ms */
+	cfg->eClockSource = ADI_TMR_CLOCK_HFOSC;
+	ret = adi_pwr_GetClockFrequency(ADI_CLOCK_HCLK, &clk_freq);
+	if (ret != ADI_PWR_SUCCESS)
+		return -ret;
+
+	/* Select prescaler */
+	aux = (float)UINT16_MAX / clk_freq * NS_PER_SEC;
+	if (period > 256 * aux)
+		return -EINVAL;
+	if (period > 64 * aux) {
+		cfg->ePrescaler = ADI_TMR_PRESCALER_256;
+		base_time = 256;
+	} else if (period > 16 * aux) {
+		cfg->ePrescaler = ADI_TMR_PRESCALER_64;
+		base_time = 64;
+	} else if (period > 1 * aux) {
+		cfg->ePrescaler = ADI_TMR_PRESCALER_16;
+		base_time = 16;
+	} else {
+		cfg->ePrescaler = ADI_TMR_PRESCALER_1;
+		if (cfg->bSyncBypass)
+			base_time = 1;
+		else
+			base_time = 0;
+	}
+	base_time = base_time * NS_PER_SEC / clk_freq;
+	cfg->nLoad = period / base_time;
+	cfg->nAsyncLoad = cfg->nLoad;
+	*match = cfg->nLoad - duty / base_time;
+
+	return SUCCESS;
+}
+
+static int32_t update_pwm_config(struct pwm_desc *desc)
+{
+	ADI_TMR_CONFIG		cfg;
+	ADI_TMR_PWM_CONFIG	pwm_cfg;
+	ADI_TMR_RESULT		ret;
+	uint32_t 		match_val;
+
+	if (!desc)
+		return -EINVAL;
+
+	cfg = (ADI_TMR_CONFIG) {
+		.bCountingUp = false,
+		.bPeriodic = true,
+		.bReloading = true,
+		.bSyncBypass = true
+	};
+	ret = config_clock_init(&cfg, desc->period_ns, desc->duty_cycle_ns,
+				&match_val);
+	if (IS_ERR_VALUE(ret))
+		return ret;
+	while (ADI_TMR_DEVICE_BUSY == adi_tmr_ConfigTimer(desc->id, &cfg));
+	if (ret != ADI_TMR_SUCCESS)
+		return -ret;
```

Comment:
ret did not change value from the last check

Reply:
I modifed the logic